firewood/src/proof.rs

```diff
@@ -455,93 +447,50 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
+            .map_err(ProofError::InvalidNode)?;
+        let addr = new_node.as_ptr();
+        match new_node.inner() {
+            NodeType::Leaf(n) => {
+                let cur_key = &n.path().0;
                 // Check if the key of current node match with the given key.
-                if key.len() < cur_key.len() || key[..cur_key.len()] != cur_key {
-                    let ext_ptr = get_ext_ptr(merkle, term, Data(data), CurKey(cur_key))?;
-
-                    return Ok((ext_ptr, None, 0));
+                if key.len() < cur_key.len() || &key[..cur_key.len()] != cur_key {
+                    return Ok((addr, None, 0));
                 }
 
-                let subproof = if term {
-                    Some(SubProof {
-                        encoded: data.clone(),
-                        hash: None,
-                    })
-                } else {
-                    generate_subproof(data.clone()).map(Some)?
-                };
-
-                let cur_key_len = cur_key.len();
-
-                let ext_ptr = get_ext_ptr(merkle, term, Data(data), CurKey(cur_key))?;
-
-                Ok((ext_ptr, subproof, cur_key_len))
+                let subproof = Some(SubProof {
+                    encoded: n.data().to_vec(),
+                    hash: None,
+                });
+                Ok((addr, subproof, cur_key.len()))
             }
+            NodeType::Extension(n) => {
+                let cur_key = &n.path().0;
 
-            BRANCH_NODE_SIZE => {
-                // we've already validated the size, that's why we can safely unwrap
-                let data = items.pop().unwrap().decode()?;
-                // Extract the value of the branch node and set to None if it's an empty Vec
-                let value = Some(data).filter(|data| !data.is_empty());
-
-                // Record encoded values of all children.
-                let mut chd_encoded: [Option<Vec<u8>>; NBRANCH] = Default::default();
-
-                // we popped the last element, so their should only be NBRANCH items left
-                for (i, chd) in items.into_iter().enumerate() {
-                    let data = chd.decode()?;
-                    chd_encoded[i] = Some(data).filter(|data| !data.is_empty());
-                }
-
-                // If the node is the last one to be decoded, then no subproof to be extracted.
-                if end_node {
-                    let branch_ptr = build_branch_ptr(merkle, value, chd_encoded)?;
-
-                    return Ok((branch_ptr, None, 1));
+                // Check if the key of current node match with the given key.
+                if key.len() < cur_key.len() || &key[..cur_key.len()] != cur_key {
+                    return Ok((addr, None, 0));
                 }
 
-                if key.is_empty() {
-                    return Err(ProofError::NoSuchNode);
-                }
+                let encoded = n.chd_encoded().ok_or(ProofError::InvalidData)?.to_vec();
```

Comment:
This allocation is unfortunate... is it possible to have a `new_node.into_inner()` such that you can remove fields from the parent?

Reply:
Will try to address it in a follow up.

Reply:
Tried it out, but it seems cannot avoid the `clone()`, as it gives `error cannot move out of a shared reference`.